internal/command/sandbox/subst.go

```diff
@@ -0,0 +1,182 @@
+				}
+				conflicts[tv.Var] = append(conflicts[tv.Var], tVal)
+				continue
+			}
+		}
+		substMap[tv.Var] = tv.Val
+	}
+	if len(conflicts) == 0 {
+		return substMap, nil
+	}
+	conflictKeys := make([]string, 0, len(conflicts))
+	for k := range conflicts {
+		conflictKeys = append(conflictKeys, k)
+	}
+	sort.Strings(conflictKeys)
+	msgs := make([]string, 0, len(conflictKeys))
+	for _, key := range conflictKeys {
+		vals := strings.Join(conflicts[key], ", ")
+		msgs = append(msgs, fmt.Sprintf("%s: %s", key, vals))
+	}
+	return nil, fmt.Errorf("conflicting variable defs:\n\t%s\n", strings.Join(msgs, "\n\t"))
+}
+
+func substTemplate(sbt *any, substMap map[string]string) error {
+	vars := map[string]struct{}{}
+	err := substTemplateRec(sbt, substMap, vars)
+	if err != nil {
+		return err
+	}
+	notExpanded := []string{}
+	for k := range vars {
+		if _, ok := substMap[k]; !ok {
+			notExpanded = append(notExpanded, k)
+		}
+	}
+	if len(notExpanded) > 0 {
+		return fmt.Errorf("unexpanded variables: %s", strings.Join(notExpanded, ", "))
+	}
+	return nil
+}
+
+func substTemplateRec(sbt *any, substMap map[string]string, vars map[string]struct{}) error {
+	switch x := (*sbt).(type) {
+	case map[string]any:
+		for k, v := range x {
+			if err := substTemplateRec(&v, substMap, vars); err != nil {
+				return err
+			}
+			x[k] = v
+		}
+
+	case []any:
+		for _, v := range x {
+			if err := substTemplateRec(&v, substMap, vars); err != nil {
+				return err
+			}
+		}
+	case string:
+		*sbt = substString(x, substMap, vars)
+	default:
+	}
+	return nil
+}
+
+func substString(s string, substMap map[string]string, vars map[string]struct{}) string {
+	matches := config.VarRefRx.FindAllStringSubmatchIndex(s, -1)
+	if matches == nil {
+		return s
+	}
+	result := []string{}
+	cur, start, end := 0, 0, 0
+	for i := range matches {
+		// begin and end of submatch corresponding to variable name
+		// in @{<var-name>}.
+		start, end = matches[i][2], matches[i][3]
+		// store any skipped string
+		if cur < start-2 {
+			result = append(result, s[cur:start-2]) // @{
+		}
+		v := s[start:end]
+		end++ // }
+		cur = end
+		vars[v] = struct{}{}
+		repl, ok := substMap[v]
+		if !ok {
+			// unsubstituted variables are handled
+			// in substTemplate to report all of them
+			// no error is reported here.
+			continue
+		}
+		result = append(result, repl)
+	}
+	if end < len(s) {
+		result = append(result, s[end:])
+	}
+	return strings.Join(result, "")
+}
+
+func unstructuredToSandbox(un any) (*models.Sandbox, error) {
+	if err := port2Int(&un); err != nil {
+		return nil, err
+	}
+	d, err := json.Marshal(un)
+	if err != nil {
+		return nil, err
+	}
+	var sb models.Sandbox
+	if err := json.Unmarshal(d, &sb); err != nil {
+		return nil, err
+	}
+	return &sb, nil
+}
+
+// translates all port values to ints if they are strings.
+func port2Int(un *any) error {
+	switch x := (*un).(type) {
+	case map[string]any:
+		for k, v := range x {
+			if k != "port" {
```

Comment:
Deciding only by the key seems brittle, even though we do control the schema here. If there isn't a simple way to do this more accurately, I'd rather start with support for only string fields and work on this later.

As an example of what I mean by "more accurately", I could imagine passing a "reference" object down through the substitution funcs and use reflection to look up the desired Go type of a given field when expanding templates in it. That would also let us continue to reject something like `"port": "123"` without any template expression, since automatic conversion in that case is surprising.

Another possibility could be to define syntax like `@{#var}` to specify that the value should be converted to a numeric type after substitution. That seems like it puts more of the burden on the user, though.

Reply:
I agree it is brittle, but 1) it is what was suggested previously without objection, and 2) it works for the functionality we have in place currently and 3) it is private and so doesn't risk being called outside the current context.  Also I believe it can easily enough be extended as needed if/when we add reference objects (and subsequently deprecated).

Inserting expansion information in the syntax, like `@{#var}`, seems to invite thinking about a world of templating with references to structured, typed data (solving this instance well is close to solving the more general problem).  This would likely would take some time for us since it's not really our focus.  At the same time, ports are often configurable, and all other scalar fields work which can lead to expectations that it should work for ports too.  So it seems to me like we have a valid use case that we can reasonably support for future extensions and the users would benefit from this in the interim.

To continue, we can sync up to discuss tomorrow if you like;  we can also ask for more opinions (@foxish @arjun-sd?) or continue here as you prefer.